Add s3 remote cache

bpaquet opened the pull request:
Largely inspired from https://github.com/moby/buildkit/pull/2323

bpaquet commented:
I'm currently testing this under load @doctolib

tonistiigi commented:
Same questions as https://github.com/moby/buildkit/pull/2323 .
- How do we test this? We don't necessarily need to run existing tests with it but we should have something that verifies the basics on the CI so we don't accidentally break it. Especially because none of the current popular downstream would have many users on it(at least initially). A separate workflow would be ok. Or if that is hard then minimally external service that tests master branch and notifies if things go wrong.
- Are you committing to maintaining this?

Why is this 40k lines bigger than the previous PR?

Take a look at the CI results as well.

cc @sethpollack

bpaquet commented:
> Same questions as #2323 .
> 
> * How do we test this? 

I'm currenlty testing this on my CI.

> We don't necessarily need to run existing tests with it but we should have something that verifies the basics on the CI so we don't accidentally break it. Especially because none of the current popular downstream would have many users on it(at least initially). A separate workflow would be ok. Or if that is hard then minimally external service that tests master branch and notifies if things go wrong.

What do you mean exactly? A dedicated workflow in github which test this? An integration test? Do you have any base I can start from?

> * Are you committing to maintaining this?

Yes.

> 
> Why is this 40k lines bigger than the previous PR?

New version of the aws s3 sdk :(

> 
> Take a look at the CI results as well.
> 
> cc @sethpollack

tonistiigi commented:
> What do you mean exactly? A dedicated workflow in github which test this? An integration test? Do you have any base I can start from?

I assume testing something like this requires either token credentials or some kind of mock service. Our regular integration tests for things like this are in `client_test.go` or `dockerfile_test.go`. But as this would likely be something that can't just be run in a container with no external dependencies it can be a special github actions workflow. Eg. we currently have special workflow for windows testing. In that workflow/test you should run the basic loop of building, exporting cache, prune local cache, rebuild with cache and verify it matched. If you want you can use external tools in that workflow, eg. you could use `buildx create` for simplicity. @crazy-max

bpaquet commented:
@tonistiigi Can you review?

tonistiigi commented on hack/s3_test/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM debian:bullseye-slim
+
+RUN apt-get update \
+  && apt-get install -y --no-install-recommends wget ca-certificates containerd
```
Why is containerd needed here?

bpaquet replied:
Buildkit does not start without it 
```
WARN[2022-05-06T08:27:58Z] skipping oci worker, as runc does not exist
WARN[2022-05-06T08:27:58Z] skipping containerd worker, as "/run/containerd/containerd.sock" does not exist
buildkitd: no worker found, rebuild the buildkit daemon?
```

May be you have a magical option to add to `buildkitd -debugaddr 0.0.0.0:8060` :)

tonistiigi commented on hack/s3_test/test1/Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM public.ecr.aws/debian/debian:bullseye-slim
+
+RUN echo first
```
These commands should write a constant plus a random value. See https://github.com/moby/buildkit/blob/master/client/client_test.go#L1410 . Then when you build them export the result to a local directory and check that the unique part has not change. Therefore cache matched.

bpaquet replied:
Done

tonistiigi reviewed (commented):
Add the ./hack/s3_test/run_test.sh invocation to https://github.com/moby/buildkit/blob/master/.github/workflows/build.yml with `needs: [base]` .

I added a commit that fixes up the Dockerfile and includes the correct buildkit binaries. 

It doesn't work for me locally though

```
Documentation: https://docs.min.io
Finished loading IAM sub-system (took 0.0s of 0.0s to load data).
+ mc alias set myminio http://172.17.0.2:9000 minioadmin minioadmin
mc: Configuration written to `/root/.mc/config.json`. Please update your access credentials.
mc: Successfully created `/root/.mc/share`.
mc: Initialized share uploads `/root/.mc/share/uploads.json` file.
mc: Initialized share downloads `/root/.mc/share/downloads.json` file.
mc: <ERROR> Unable to initialize new alias from the provided credentials. Get "http://172.17.0.2:9000/probe-bucket-sign-tszmiipmum4w/?location=": dial tcp 172.17.0.2:9000: connect: connection refused.
```

Assuming this works it doesn't actually look as complicated as I would have assumed. You can also consider adding the `minio` to the main Dockerfile `integration-tests` stage and adding a regular Go integration test.

FYI @crazy-max

tonistiigi commented:
> Not sure how to taunch them in GHA, especially with the --privileged.

`--privileged` is not a problem in GH actions.

> I duplicated your readerat.go from the go-actions-cache. Should we move it in buildkit repo?

Need to think about this later. I don't want an import cycle there.

crazy-max commented:
> You can also consider adding the `minio` to the main Dockerfile `integration-tests` stage and adding a regular Go integration test.

Added a commit to add minio to integration tests stage and also use it through linked context in the bake definition for s3 test.

I will add another commit to create a new job in our GHA workflow.

bpaquet commented:
> Add the ./hack/s3_test/run_test.sh invocation to https://github.com/moby/buildkit/blob/master/.github/workflows/build.yml with `needs: [base]` .
> 
> I added a commit that fixes up the Dockerfile and includes the correct buildkit binaries.
> 
> It doesn't work for me locally though
> 
> ```
> Documentation: https://docs.min.io
> Finished loading IAM sub-system (took 0.0s of 0.0s to load data).
> + mc alias set myminio http://172.17.0.2:9000 minioadmin minioadmin
> mc: Configuration written to `/root/.mc/config.json`. Please update your access credentials.
> mc: Successfully created `/root/.mc/share`.
> mc: Initialized share uploads `/root/.mc/share/uploads.json` file.
> mc: Initialized share downloads `/root/.mc/share/downloads.json` file.
> mc: <ERROR> Unable to initialize new alias from the provided credentials. Get "http://172.17.0.2:9000/probe-bucket-sign-tszmiipmum4w/?location=": dial tcp 172.17.0.2:9000: connect: connection refused.
> ```
> 

Fixed
> Assuming this works it doesn't actually look as complicated as I would have assumed. You can also consider adding the `minio` to the main Dockerfile `integration-tests` stage and adding a regular Go integration test.

I prefer to keep it like this if you dont mind.

> 
> FYI @crazy-max

I

bpaquet commented:
> > You can also consider adding the `minio` to the main Dockerfile `integration-tests` stage and adding a regular Go integration test.
> 
> Added a commit to add minio to integration tests stage and also use it through linked context in the bake definition for s3 test.
> 
> I will add another commit to create a new job in our GHA workflow.

@crazy-max just force pushed, sorry may be I drop some stuff. I added something in the build workflow, please feel free to refact.

crazy-max commented:
> @crazy-max just force pushed, sorry may be I drop some stuff. I added something in the build workflow, please feel free to refact.

Yes just saw that, that's fine. I bring it back.

> I prefer to keep it like this if you dont mind.

In the future I think we don't want to rely on scripting but Go integration tests like we do for containerd and dockerd. See https://github.com/moby/buildkit/tree/master/util/testutil/integration

tonistiigi commented on cache/remotecache/s3/s3.go:
```diff
@@ -0,0 +1,541 @@
+}
+
+func (e *exporter) Config() remotecache.Config {
+	return remotecache.Config{}
```
This should return `Compression: default`

bpaquet replied:
done

tonistiigi commented on cache/remotecache/s3/s3.go:
```diff
@@ -0,0 +1,541 @@
+	return config.Prefix + config.BlobsPrefix + dgst.String()
+}
+
+type cache struct {
```
Calling this `cache` seems kind of confusing. It's more like `s3Client`

bpaquet replied:
Done

tonistiigi commented on cache/remotecache/s3/s3.go:
```diff
@@ -0,0 +1,541 @@
+}
+
+func (c *cache) getManifest(key string, config *v1.CacheConfig) (bool, error) {
+	fmt.Printf("Start downloading s3://%s/%s\n", c.config.Bucket, key)
```
Writing to stdout is not allowed in here. If you want debug logs then you need to write them to the context logger.

bpaquet replied:
removed all debug

tonistiigi commented on cache/remotecache/s3/s3.go:
```diff
@@ -0,0 +1,541 @@
+}
+
+func isNotFound(err error) bool {
+	awsErr, ok := err.(awserr.Error)
```
This should use `errors.As`

bpaquet replied:
Done

tonistiigi commented on hack/s3_test/test1/Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM public.ecr.aws/debian/debian:bullseye-slim
+
+RUN cat /dev/urandom | head -c 100 | sha256sum > unique_first
+RUN cat /dev/urandom | head -c 100 | sha256sum > unique_second
```
Add a another stage in here

```
FROM scratch
COPY --link --from=build /unique_first /
COPY --link --from=build /unique_second /
```

And in the test add `--output destdir` . After cache import the contents of the files need to remain the same.

bpaquet replied:
Done

tonistiigi reviewed (commented):
LGTM but try to squash the commits a bit. It doesn't need to be a single commit but we should avoid commits that just say "fixes" or "code review". Maybe squash these 2 with the first commit. The others can remain as they are.

tonistiigi commented on cache/remotecache/s3/s3.go:
```diff
@@ -0,0 +1,493 @@
+func news3Client(config Config) (*s3Client, error) {
+	s, err := sess.NewSession(&aws.Config{Region: &config.Region, Endpoint: &config.EndpointURL, S3ForcePathStyle: &config.S3ForcePathStyle})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create session: %w", err)
```
Use `errors.Errorf` always for stacktraces

tonistiigi commented on hack/s3_test/run_test.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh -ex
```
This file looks unused now

bpaquet replied:
I'm using it to launch the test locally.

tonistiigi replied:
If we have a helper script we should call it in the CI. Otherwise nobody notices if it breaks or starts to behave differently than CI invocations.

bpaquet replied:
Used it in the CI

crazy-max commented on go.mod:
```diff
@@ -76,7 +80,22 @@ require (
 	google.golang.org/grpc v1.45.0
 )
 
+require github.com/aws/aws-sdk-go-v2/config v1.15.5
```
should be moved in the previous require block

bpaquet replied:
Fixed. I should do something wrong when I update mod / vendor :(

crazy-max reviewed (approved):
LGTM

in a follow-up we should think about proper integration tests in testutil but might needs to split first as suggested by @tonistiigi in https://github.com/moby/buildkit/issues/2734#issuecomment-1124422177

also README starts to be quite huge. wonder if we should not move some sections in dedicated pages under `./docs` like the cache exporters.

zchee commented:
@tonistiigi Can I send support GCS as well?

bpaquet commented:
I’m using it to run the test locally.

On Fri 13 May 2022 at 00:03, Tõnis Tiigi ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In hack/s3_test/run_test.sh
> <https://github.com/moby/buildkit/pull/2824#discussion_r871842760>:
>
> > @@ -0,0 +1,7 @@
> +#!/bin/sh -ex
>
> This file looks unused now
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/buildkit/pull/2824#pullrequestreview-971502453>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACSJXKMF2JOLXFTVW7POXTVJV53HANCNFSM5UJLRGCA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
